<center>
    <h1>
        <b>Mario Ferreyra | Emiliano Kokic</b>
    </h1>
    <h4>Diplomatura AACSyA 2018 - FaMAF - UNC</h4>
    <h3>Análisis Exploratorio y Curación de Datos</h3>
</center>

# **Laboratorio: Importando los Datos (Dataset)**

In [1]:
%matplotlib inline
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from IPython.display import display

In [2]:
pd.__version__

'0.23.0'

## 1. Importando los Datos (Dataset)

### 1.1 Verificamos que no hay problemas en la importación

In [3]:
building_permits = pd.read_csv("Building_Permits.csv")

/home/mferreyra/.virtualenvs/diplodatos-ayc/lib/python3.6/site-packages/IPython/core/interactiveshell.py:2785: DtypeWarning: Columns (22,32) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [4]:
building_permits.shape

(198900, 43)

In [5]:
building_permits.dropna().shape

(0, 43)

In [6]:
building_permits.shape

(198900, 43)

In [7]:
building_permits

Permit Number  Permit Type            Permit Type Definition  \
0       201505065519            4                      sign - erect   
1       201604195146            4                      sign - erect   
2       201605278609            3  additions alterations or repairs   
3       201611072166            8            otc alterations permit   
4       201611283529            6                       demolitions   
5       201706149344            8            otc alterations permit   
6       201706300814            8            otc alterations permit   
7            M803667            8            otc alterations permit   
8            M804227            8            otc alterations permit   
9            M804767            8            otc alterations permit   
10           M805287            8            otc alterations permit   
11           M805907            8            otc alterations permit   
12           M806447            8            otc alterations permit   
13      201707252846            8            otc alterations permit   
14           M813729            8            otc alterations permit   
15           M813907            8            otc alterations permit   
16           M813967            8            otc alterations permit   
17           M814148            8            otc alterations permit   
18      201707283165            8            otc alterations permit   
19           M814368            8            otc alterations permit   
20           M814967            8            otc alterations permit   
21      201708013476            8            otc alterations permit   
22      201708013544            3  additions alterations or repairs   
23           M816927            8            otc alterations permit   
24      201708084249            8            otc alterations permit   
25      201708165014            8            otc alterations permit   
26           M820728            8            otc alterations permit   
27           M821207            8            otc alterations permit   
28           M821268            8            otc alterations permit   
29           M821847            8            otc alterations permit   
...              ...          ...                               ...   
198870       M893328            8            otc alterations permit   
198871       M893347            8            otc alterations permit   
198872  201802232165            8            otc alterations permit   
198873  201802232166            8            otc alterations permit   
198874  201802232106            8            otc alterations permit   
198875  201802232167            8            otc alterations permit   
198876       M893367            8            otc alterations permit   
198877  201802232169            8            otc alterations permit   
198878  201802232170            8            otc alterations permit   
198879  201802232171            8            otc alterations permit   
198880       M893387            8            otc alterations permit   
198881       M893407            8            otc alterations permit   
198882  201802232172            8            otc alterations permit   
198883       M893427            8            otc alterations permit   
198884       M893447            8            otc alterations permit   
198885  201802232173            8            otc alterations permit   
198886  201709066971            3  additions alterations or repairs   
198887  201710040362            8            otc alterations permit   
198888  201710110995            8            otc alterations permit   
198889  201710242114            8            otc alterations permit   
198890  201711274822            3  additions alterations or repairs   
198891  201711274844            8            otc alterations permit   
198892  201712045380            8            otc alterations permit   
198893  201712055472            8            otc alterations permit   
198894  201712055493           

In [8]:
# Set seed for reproducibility
np.random.seed(0)

In [9]:
building_permits.sample(5)

Permit Number  Permit Type            Permit Type Definition  \
40553   201403039652            8            otc alterations permit   
169731  201510159735            3  additions alterations or repairs   
19180        M409787            8            otc alterations permit   
68047   201411191888            8            otc alterations permit   
64238        M527228            8            otc alterations permit   

       Permit Creation Date Block  Lot  Street Number Street Number Suffix  \
40553            03/03/2014  3732  008            400                  NaN   
169731           10/15/2015  2609  028             79                  NaN   
19180            07/22/2013  4624  031            178                  NaN   
68047            11/19/2014  0039  109            294                  NaN   
64238            10/14/2014  1251  002            707                  NaN   

        Street Name Street Suffix      ...        Existing Construction Type  \
40553    Clementina            St      ...                               NaN   
169731  Buena Vista            Tr      ...                               5.0   
19180    West Point            Rd      ...                               NaN   
68047     Francisco            St      ...                               5.0   
64238          Cole            St      ...                               NaN   

       Existing Construction Type Description Proposed Construction Type  \
40553                                     NaN                        1.0   
169731                         wood frame (5)                        5.0   
19180                                     NaN                        NaN   
68047                          wood frame (5)                        5.0   
64238                                     NaN                        NaN   

       Proposed Construction Type Description Site Permit Supervisor District  \
40553                           constr type 1         NaN                 6.0   
169731                         wood frame (5)         NaN                 8.0   
19180                                     NaN         NaN                10.0   
68047                          wood frame (5)         NaN                 3.0   
64238                                     NaN         NaN                 5.0   

       Neighborhoods - Analysis Boundaries  Zipcode  \
40553                      South of Market  94103.0   
169731                 Castro/Upper Market  94117.0   
19180                Bayview Hunters Point  94124.0   
68047                          North Beach  94133.0   
64238                       Haight Ashbury  94117.0   

                                         Location      Record ID  
40553   (37.780460571778164, -122.40450626524974)  1334094491645  
169731   (37.76757916496494, -122.43793170417105)  1399356139170  
19180    (37.73524725436046, -122.38063828309745)  1311685491725  
68047   (37.805257822817126, -122.40998545760392)  1362881288870  
64238    (37.76836885973765, -122.45074431487859)   135886493776  

[5 rows x 43 columns]

Veamos si la descripción del dataset se corresponde con lo levantado por Pandas:

In [10]:
rows = [["Permit Number", "-", "String"],
        ["Permit Type", "-", "Numeric"],
        ["Permit Type Definition", "-", "String"],
        ["Permit Creation Date", "-", "DateTime"],
        ["Block", "-", "String"],
        ["Lot", "-", "String"],
        ["Street Number", "-", "Numeric"],
        ["Street Number Suffix", "-", "String"],
        ["Street Name", "-", "String"],
        ["Street Suffix", "-", "String"],
        ["Unit", "-", "String"],
        ["Unit Suffix", "-", "String"],
        ["Description", "-", "String"],
        ["Current Status", "-", "String"],
        ["Current Status Date", "-", "DateTime"],
        ["Filed Date", "-", "DateTime"],
        ["Issued Date", "-", "String"],
        ["Completed Date", "-", "String"],
        ["First Construction Document Date", "-", "String"],
        ["Structural Notification", "-", "String"],
        ["Number of Existing Stories", "-", "String"],
        ["Number of Proposed Stories", "-", "String"],
        ["Voluntary Soft-Story Retrofit", "-", "String"],
        ["Fire Only Permit", "-", "String"],
        ["Permit Expiration Date", "-", "String"],
        ["Estimated Cost", "-", "String"],
        ["Revised Cost", "-", "String"],
        ["Existing Use", "-", "String"],
        ["Existing Units", "-", "String"],
        ["Proposed Use", "-", "String"],
        ["Proposed Units", "-", "String"],
        ["Plansets", "-", "String"],
        ["TIDF Compliance", "-", "String"],
        ["Existing Construction Type", "-", "String"],
        ["Existing Construction Type Description", "-", "String"],
        ["Proposed Construction Type", "-", "String"],
        ["Proposed Construction Type Description", "-", "String"],
        ["Site Permit", "-", "String"],
        ["Supervisor District", "-", "Numeric"],
        ["Neighborhoods - Analysis Boundaries", "-", "String"],
        ["Zipcode", "-", "Numeric"],
        ["Location", "-", "String"],
        ["Record ID", "-", "Numeric"]]

pd.DataFrame(rows, columns=["Field name", "Field description", "Type"])

Field name Field description      Type
0                            Permit Number                 -    String
1                              Permit Type                 -   Numeric
2                   Permit Type Definition                 -    String
3                     Permit Creation Date                 -  DateTime
4                                    Block                 -    String
5                                      Lot                 -    String
6                            Street Number                 -   Numeric
7                     Street Number Suffix                 -    String
8                              Street Name                 -    String
9                            Street Suffix                 -    String
10                                    Unit                 -    String
11                             Unit Suffix                 -    String
12                             Description                 -    String
13                          Current Status                 -    String
14                     Current Status Date                 -  DateTime
15                              Filed Date                 -  DateTime
16                             Issued Date                 -    String
17                          Completed Date                 -    String
18        First Construction Document Date                 -    String
19                 Structural Notification                 -    String
20              Number of Existing Stories                 -    String
21              Number of Proposed Stories                 -    String
22           Voluntary Soft-Story Retrofit                 -    String
23                        Fire Only Permit                 -    String
24                  Permit Expiration Date                 -    String
25                          Estimated Cost                 -    String
26                            Revised Cost                 -    String
27                            Existing Use                 -    String
28                          Existing Units                 -    String
29                            Proposed Use                 -    String
30                          Proposed Units                 -    String
31                                Plansets                 -    String
32                         TIDF Compliance                 -    String
33              Existing Construction Type                 -    String
34  Existing Construction Type Description                 -    String
35              Proposed Construction Type                 -    String
36  Proposed Construction Type Description                 -    String
37                             Site Permit                 -    String
38                     Supervisor District                 -   Numeric
39     Neighborhoods - Analysis Boundaries                 -    String
40                                 Zipcode                 -   Numeric
41                                Location                 -    String
42                               Record ID                 -   Numeric

In [12]:
building_permits.dtypes

Permit Number                              object
Permit Type                                 int64
Permit Type Definition                     object
Permit Creation Date                       object
Block                                      object
Lot                                        object
Street Number                               int64
Street Number Suffix                       object
Street Name                                object
Street Suffix                              object
Unit                                      float64
Unit Suffix                                object
Description                                object
Current Status                             object
Current Status Date                        object
Filed Date                                 object
Issued Date                                object
Completed Date                             object
First Construction Document Date           object
Structural Notification                    object


Vemos que no reconocio las fechas **Permit Creation Date**, **Current Status Date**, **Filed Date**

In [13]:
building_permits.describe()

Permit Type  Street Number          Unit  Number of Existing Stories  \
count  198900.000000  198900.000000  29479.000000               156116.000000   
mean        7.522323    1121.728944     78.517182                    5.705773   
std         1.457451    1135.768948    326.981324                    8.613455   
min         1.000000       0.000000      0.000000                    0.000000   
25%         8.000000     235.000000      0.000000                    2.000000   
50%         8.000000     710.000000      0.000000                    3.000000   
75%         8.000000    1700.000000      1.000000                    4.000000   
max         8.000000    8400.000000   6004.000000                   78.000000   

       Number of Proposed Stories  Estimated Cost  Revised Cost  \
count               156032.000000    1.608340e+05  1.928340e+05   
mean                     5.745043    1.689554e+05  1.328562e+05   
std                      8.613284    3.630386e+06  3.584903e+06   
min                      0.000000    1.000000e+00  0.000000e+00   
25%                      2.000000    3.300000e+03  1.000000e+00   
50%                      3.000000    1.100000e+04  7.000000e+03   
75%                      4.000000    3.500000e+04  2.870750e+04   
max                     78.000000    5.379586e+08  7.805000e+08   

       Existing Units  Proposed Units       Plansets  \
count   147362.000000   147989.000000  161591.000000   
mean        15.666164       16.510950       1.274650   
std         74.476321       75.220444      22.407345   
min          0.000000        0.000000       0.000000   
25%          1.000000        1.000000       0.000000   
50%          1.000000        2.000000       2.000000   
75%          4.000000        4.000000       2.000000   
max       1907.000000     1911.000000    9000.000000   

       Existing Construction Type  Proposed Construction Type  \
count               155534.000000               155738.000000   
mean                     4.072878                    4.089529   
std                      1.585756                    1.578766   
min                      1.000000                    1.000000   
25%                      3.000000                    3.000000   
50%                      5.000000                    5.000000   
75%                      5.000000                    5.000000   
max                      5.000000                    5.000000   

       Supervisor District        Zipcode     Record ID  
count        197183.000000  197184.000000  1.989000e+05  
mean              5.538403   94115.500558  1.162048e+12  
std               2.887041       9.270131  4.918215e+11  
min               1.000000   94102.000000  1.293532e+10  
25%               3.000000   94109.000000  1.308567e+12  
50%               6.000000   94114.000000  1.371840e+12  
75%               8.000000   94122.000000  1.435000e+12  
max              11.000000   94158.000000  1.498342e+12

In [14]:
building_permits.describe(include=['object'])

Permit Number  Permit Type Definition Permit Creation Date   Block  \
count         198900                  198900               198900  198900   
unique        181495                       8                 1291    4896   
top     201602179765  otc alterations permit           09/15/2017    3708   
freq             101                  178844                  413    1195   

           Lot Street Number Suffix Street Name Street Suffix Unit Suffix  \
count   198900                 2216      198900        196132        1961   
unique    1055                   18        1704            21         164   
top        001                    A      Market            St           A   
freq     10114                 1501        5443        138358         346   

         Description                   ...                     \
count         198610                   ...                      
unique        134272                   ...                      
top     street space                   ...                      
freq           22163                   ...                      

       Fire Only Permit Permit Expiration Date       Existing Use  \
count             18827                 147020             157786   
unique                1                   2232                 93   
top                   Y             02/28/2018  1 family dwelling   
freq              18827                    302              46766   

             Proposed Use TIDF Compliance  \
count              156461               2   
unique                 94               2   
top     1 family dwelling               P   
freq                46346               1   

       Existing Construction Type Description  \
count                                  155534   
unique                                      5   
top                            wood frame (5)   
freq                                   113350   

       Proposed Construction Type Description Site Permit  \
count                                  155738        5359   
unique                                      5           1   
top                            wood frame (5)           Y   
freq                                   114382        5359   

       Neighborhoods - Analysis Boundaries  \
count                               197175   
unique                                  41   
top         Financial District/South Beach   
freq                                 21816   

                                       Location  
count                                    197200  
unique                                    57604  
top     (37.79226164705184, -122.4034859571375)  
freq                                        554  

[4 rows x 28 columns]

### Habilitamos chequeos al importar los datos

Como por ejemplo las fechas

In [15]:
display(building_permits['Permit Creation Date'].min())
display(building_permits['Permit Creation Date'].max())

'01/02/2013'

'12/31/2015'

In [16]:
building_permits['Permit Creation Date'].max() - building_permits['Permit Creation Date'].min()

TypeError: unsupported operand type(s) for -: 'str' and 'str'

Indiquemos que las columnas **Permit Creation Date**, **Current Status Date**, **Filed Date** son fechas como indica su descripcion original.

In [17]:
building_permits = pd.read_csv("Building_Permits.csv",
                               parse_dates=["Permit Creation Date", "Current Status Date", "Filed Date"])

/home/mferreyra/.virtualenvs/diplodatos-ayc/lib/python3.6/site-packages/IPython/core/interactiveshell.py:2785: DtypeWarning: Columns (22,32) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [18]:
building_permits.dtypes

Permit Number                                     object
Permit Type                                        int64
Permit Type Definition                            object
Permit Creation Date                      datetime64[ns]
Block                                             object
Lot                                               object
Street Number                                      int64
Street Number Suffix                              object
Street Name                                       object
Street Suffix                                     object
Unit                                             float64
Unit Suffix                                       object
Description                                       object
Current Status                                    object
Current Status Date                       datetime64[ns]
Filed Date                                datetime64[ns]
Issued Date                                       object
Completed Date                 

In [19]:
building_permits.sample(5)

Permit Number  Permit Type            Permit Type Definition  \
60635   201409105988            3  additions alterations or repairs   
110430  201601046309            8            otc alterations permit   
165042  201707212672            3  additions alterations or repairs   
79200   201503171100            8            otc alterations permit   
76487   201502208860            8            otc alterations permit   

       Permit Creation Date Block   Lot  Street Number Street Number Suffix  \
60635            2014-09-10  0479   027           1356                  NaN   
110430           2016-01-04  1276  001H            202                  NaN   
165042           2017-07-21  3622   004            818                  NaN   
79200            2015-03-17  1509   009            547                  NaN   
76487            2015-02-20  1923   023           1662                  NaN   

       Street Name Street Suffix      ...        Existing Construction Type  \
60635     Chestnut            St      ...                               5.0   
110430     Grattan            St      ...                               5.0   
165042         Noe            St      ...                               5.0   
79200         35th            Av      ...                               5.0   
76487         21st            Av      ...                               5.0   

       Existing Construction Type Description Proposed Construction Type  \
60635                          wood frame (5)                        5.0   
110430                         wood frame (5)                        5.0   
165042                         wood frame (5)                        5.0   
79200                          wood frame (5)                        5.0   
76487                          wood frame (5)                        5.0   

       Proposed Construction Type Description Site Permit Supervisor District  \
60635                          wood frame (5)           Y                 2.0   
110430                         wood frame (5)         NaN                 5.0   
165042                         wood frame (5)           Y                 8.0   
79200                          wood frame (5)         NaN                 1.0   
76487                          wood frame (5)         NaN                 4.0   

       Neighborhoods - Analysis Boundaries  Zipcode  \
60635                               Marina  94123.0   
110430                      Haight Ashbury  94117.0   
165042                          Noe Valley  94114.0   
79200                       Outer Richmond  94121.0   
76487                      Sunset/Parkside  94122.0   

                                         Location      Record ID  
60635   (37.802444119694016, -122.42570568145257)  1355098392583  
110430    (37.7639230255061, -122.45133897567636)  1408159479509  
165042  (37.755878042652135, -122.43257973156736)  1471545162735  
79200   (37.778632168486645, -122.49570921682907)  1374631102139  
76487   (37.756722563984916, -122.47846371493331)  1371767120653  

[5 rows x 43 columns]

In [20]:
building_permits.describe(include='all')

Permit Number    Permit Type  Permit Type Definition  \
count         198900  198900.000000                  198900   
unique        181495            NaN                       8   
top     201602179765            NaN  otc alterations permit   
freq             101            NaN                  178844   
first            NaN            NaN                     NaN   
last             NaN            NaN                     NaN   
mean             NaN       7.522323                     NaN   
std              NaN       1.457451                     NaN   
min              NaN       1.000000                     NaN   
25%              NaN       8.000000                     NaN   
50%              NaN       8.000000                     NaN   
75%              NaN       8.000000                     NaN   
max              NaN       8.000000                     NaN   

       Permit Creation Date   Block     Lot  Street Number  \
count                198900  198900  198900  198900.000000   
unique                 1291    4896    1055            NaN   
top     2017-09-15 00:00:00    3708     001            NaN   
freq                    413    1195   10114            NaN   
first   2012-03-28 00:00:00     NaN     NaN            NaN   
last    2018-02-23 00:00:00     NaN     NaN            NaN   
mean                    NaN     NaN     NaN    1121.728944   
std                     NaN     NaN     NaN    1135.768948   
min                     NaN     NaN     NaN       0.000000   
25%                     NaN     NaN     NaN     235.000000   
50%                     NaN     NaN     NaN     710.000000   
75%                     NaN     NaN     NaN    1700.000000   
max                     NaN     NaN     NaN    8400.000000   

       Street Number Suffix Street Name Street Suffix      ...       \
count                  2216      198900        196132      ...        
unique                   18        1704            21      ...        
top                       A      Market            St      ...        
freq                   1501        5443        138358      ...        
first                   NaN         NaN           NaN      ...        
last                    NaN         NaN           NaN      ...        
mean                    NaN         NaN           NaN      ...        
std                     NaN         NaN           NaN      ...        
min                     NaN         NaN           NaN      ...        
25%                     NaN         NaN           NaN      ...        
50%                     NaN         NaN           NaN      ...        
75%                     NaN         NaN           NaN      ...        
max                     NaN         NaN           NaN      ...        

        Existing Construction Type Existing Construction Type Description  \
count                155534.000000                                 155534   
unique                         NaN                                      5   
top                            NaN                         wood frame (5)   
freq                           NaN                                 113350   
first                          NaN                                    NaN   
last                           NaN                                    NaN   
mean                      4.072878                                    NaN   
std                       1.585756                                    NaN   
min                       1.000000                                    NaN   
25%                       3.000000                                    NaN   
50%                       5.000000                                    NaN   
75%                       5.000000                                    NaN   
max                       5.000000                                    NaN   

       Proposed Construction Type Proposed Construction Type Description  \
count               155738.000000                                 155738   
unique                       

In [21]:
building_permits['Permit Creation Date'].max() - building_permits['Permit Creation Date'].min()

Timedelta('2158 days 00:00:00')

Vamos a usar como index la columna de **Permit Number**

In [22]:
building_permits = pd.read_csv("Building_Permits.csv",
                               index_col="Permit Number",
                               parse_dates=["Permit Creation Date", "Current Status Date", "Filed Date"])

/home/mferreyra/.virtualenvs/diplodatos-ayc/lib/python3.6/site-packages/IPython/core/interactiveshell.py:2785: DtypeWarning: Columns (22,32) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [23]:
building_permits

Permit Type            Permit Type Definition  \
Permit Number                                                  
201505065519             4                      sign - erect   
201604195146             4                      sign - erect   
201605278609             3  additions alterations or repairs   
201611072166             8            otc alterations permit   
201611283529             6                       demolitions   
201706149344             8            otc alterations permit   
201706300814             8            otc alterations permit   
M803667                  8            otc alterations permit   
M804227                  8            otc alterations permit   
M804767                  8            otc alterations permit   
M805287                  8            otc alterations permit   
M805907                  8            otc alterations permit   
M806447                  8            otc alterations permit   
201707252846             8            otc alterations permit   
M813729                  8            otc alterations permit   
M813907                  8            otc alterations permit   
M813967                  8            otc alterations permit   
M814148                  8            otc alterations permit   
201707283165             8            otc alterations permit   
M814368                  8            otc alterations permit   
M814967                  8            otc alterations permit   
201708013476             8            otc alterations permit   
201708013544             3  additions alterations or repairs   
M816927                  8            otc alterations permit   
201708084249             8            otc alterations permit   
201708165014             8            otc alterations permit   
M820728                  8            otc alterations permit   
M821207                  8            otc alterations permit   
M821268                  8            otc alterations permit   
M821847                  8            otc alterations permit   
...                    ...                               ...   
M893328                  8            otc alterations permit   
M893347                  8            otc alterations permit   
201802232165             8            otc alterations permit   
201802232166             8            otc alterations permit   
201802232106             8            otc alterations permit   
201802232167             8            otc alterations permit   
M893367                  8            otc alterations permit   
201802232169             8            otc alterations permit   
201802232170             8            otc alterations permit   
201802232171             8            otc alterations permit   
M893387                  8            otc alterations permit   
M893407                  8            otc alterations permit   
201802232172             8            otc alterations permit   
M893427                  8            otc alterations permit   
M893447                  8            otc alterations permit   
201802232173             8            otc alterations permit   
201709066971             3  additions alterations or repairs   
201710040362             8            otc alterations permit   
201710110995             8            otc alterations permit   
201710242114             8            otc alterations permit   
201711274822             3  additions alterations or repairs   
201711274844             8            otc alterations permit   
201712045380             8            otc alterations permit   
201712055472             8            otc alterations permit   
201712055493             8            otc alterations permit   
M862628                  8            otc alterations permit   
201712055595             8            otc alterations permit   
M863507                  8            otc alterations permit   
M863747                  8            otc alterations permit   
M864287                  8            otc alter

### 1.2 Aseguremonos de tener ids/claves únicas

#### Chequemos que no hay datos duplicados

In [24]:
building_permits[building_permits.duplicated()]

Empty DataFrame
Columns: [Permit Type, Permit Type Definition, Permit Creation Date, Block, Lot, Street Number, Street Number Suffix, Street Name, Street Suffix, Unit, Unit Suffix, Description, Current Status, Current Status Date, Filed Date, Issued Date, Completed Date, First Construction Document Date, Structural Notification, Number of Existing Stories, Number of Proposed Stories, Voluntary Soft-Story Retrofit, Fire Only Permit, Permit Expiration Date, Estimated Cost, Revised Cost, Existing Use, Existing Units, Proposed Use, Proposed Units, Plansets, TIDF Compliance, Existing Construction Type, Existing Construction Type Description, Proposed Construction Type, Proposed Construction Type Description, Site Permit, Supervisor District, Neighborhoods - Analysis Boundaries, Zipcode, Location, Record ID]
Index: []

[0 rows x 42 columns]

In [25]:
from collections import Counter

print(Counter(building_permits.index.duplicated()))
building_permits[building_permits.index.duplicated()]

Counter({False: 181495, True: 17405})


Permit Type            Permit Type Definition  \
Permit Number                                                  
201301027072             8            otc alterations permit   
201301027079             8            otc alterations permit   
201301027081             8            otc alterations permit   
201301027093             8            otc alterations permit   
201301027095             8            otc alterations permit   
201301027105             8            otc alterations permit   
M364508                  8            otc alterations permit   
M364568                  8            otc alterations permit   
201301037165             8            otc alterations permit   
201301037171             8            otc alterations permit   
201301037188             8            otc alterations permit   
201301037191             8            otc alterations permit   
201301037195             8            otc alterations permit   
201301037195             8            otc alterations permit   
201301037195             8            otc alterations permit   
201301047213             8            otc alterations permit   
201301047218             8            otc alterations permit   
201301047225             8            otc alterations permit   
201301047230             8            otc alterations permit   
M364707                  8            otc alterations permit   
201301047244             8            otc alterations permit   
201301047258             8            otc alterations permit   
201301047276             8            otc alterations permit   
201301047277             8            otc alterations permit   
201301047278             8            otc alterations permit   
201301037207             8            otc alterations permit   
201301047292             8            otc alterations permit   
201301047296             8            otc alterations permit   
201301047297             8            otc alterations permit   
201301047304             8            otc alterations permit   
...                    ...                               ...   
201802222020             8            otc alterations permit   
201802222022             8            otc alterations permit   
201802222028             8            otc alterations permit   
201802222039             8            otc alterations permit   
201802222040             8            otc alterations permit   
201802222042             8            otc alterations permit   
201802222043             8            otc alterations permit   
201802222045             8            otc alterations permit   
201802222046             8            otc alterations permit   
201802232062             8            otc alterations permit   
M892827                  8            otc alterations permit   
201802232077             8            otc alterations permit   
201802232088             3  additions alterations or repairs   
201802232097             8            otc alterations permit   
201802232097             8            otc alterations permit   
201802232110             8            otc alterations permit   
201711284887             8            otc alterations permit   
201802232115             8            otc alterations permit   
201802232126             8            otc alterations permit   
201802232126             8            otc alterations permit   
201802232131             3  additions alterations or repairs   
201802232131             3  additions alterations or repairs   
201802232137             3  additions alterations or repairs   
201504153737             3  additions alterations or repairs   
201504153735             2       new construction wood frame   
M893328                  8            otc alterations permit   
201711274822             3  additions alterations or repairs   
201711274844             8            otc alterations permit   
201712055493             8            otc alterations permit   
M862628                  8            otc alter

Vemos que tenemos 17405 duplicados (usando como Index **Permit Number**), es decir, un 8.75% del dataset original.

Por lo que nos quedamos con un 91.25% del dataset, es decir, 181495 registros.

### 1.3 Despersonalizamos los datos y los guardarmos en un nuevo archivo

In [26]:
building_permits_without_duplicates = building_permits[~building_permits.index.duplicated()]

In [27]:
display(building_permits_without_duplicates.shape)

(181495, 42)

Vamos despersonalizar las siguientes columnas:
* Street Number
* Street Name
* Zipcode
* Location

Ya que a nuestro parecer son las más relevantes a la hora de ocultar su información.

Las tecnicas que aplicaremos seran las siguientes:
* Bucket values based on custom size ranges (`Street Number`)
* Pseudonymization by replacing input value with cryptographic hash (`Street Name`)
* Mask with character (`Zipcode`)
* Replace with infoType (`Location`)


Para `Street Number` reemplazaremos los valores por `LOW`, `MEDIUM`, `HIGH`:
* Si el `Street Number` esta entre 0 y 3333 $\longrightarrow$ `LOW`
* Si el `Street Number` esta entre 3334 y 6666 $\longrightarrow$ `MEDIUM`
* Si el `Street Number` es mayor a 6666 $\longrightarrow$ `HIGH`

Para `Street Name` aplicamos MD5 sobre los nombres para obtener un valor hexadecimal.

Para `Zipcode` reemplazamos los numberos a partir del 3 digito por `*` con el fin de ocultar parte del codigo postal.

Para `Location` reemplazamos su par (latitud, longitud) por el nombre del tipo de dato que corresponde en el dataset: `String`.

In [28]:
# Bucket values based on custom size ranges
def bucketing(street_number):
    if 0 <= street_number <= 3333:
        return "LOW"
    elif 3333 < street_number <= 6666:
        return "MEDIUM"
    else:
        return "HIGH"

In [29]:
from hashlib import md5

# Pseudonymization by replacing input value with cryptographic hash
def hashit(val):
    return md5(val.encode('utf-8')).hexdigest()

In [30]:
# Mask with character
def hide_zipcode(zipcode):
    try:
        zipcode = int(zipcode)
    except:
        return zipcode

    new_zipcode = ""
    for index, c in enumerate(str(zipcode)):
        if index >= 2:
            new_zipcode += "*"
            continue
        new_zipcode += c
    
    return new_zipcode

In [31]:
# Replace with infoType
def replace_by_infotype(location):
    infotype_location = "String"
    return infotype_location

In [32]:
building_permits_without_duplicates['Street Number'].apply(bucketing).head(10)

Permit Number
201505065519    LOW
201604195146    LOW
201605278609    LOW
201611072166    LOW
201611283529    LOW
201706149344    LOW
201706300814    LOW
M803667         LOW
M804227         LOW
M804767         LOW
Name: Street Number, dtype: object

In [33]:
building_permits_without_duplicates['Street Name'].apply(hashit).head(10)

Permit Number
201505065519    58da3cd31090edfa3b22657b6faf132c
201604195146    bd03b3de4a4d14f6e1d7dc15a88b464f
201605278609    501a92c9b793cb44dfbfe0ca9ffee563
201611072166    501a92c9b793cb44dfbfe0ca9ffee563
201611283529    31840a66a8d6d223e5b0540138768838
201706149344    c5376d00573c5858adf7ed894d0491b9
201706300814    6efa70900c0db0ee83dd0a7b706596ef
M803667         4a468371d3159b132061b5830764e56e
M804227         ae8419cd44d2884fe51dae4f6d89fe1e
M804767         45a94c5ae2df00c49cca09d7d26c06ec
Name: Street Name, dtype: object

In [34]:
building_permits_without_duplicates["Zipcode"].apply(hide_zipcode).head(10)

Permit Number
201505065519    94***
201604195146    94***
201605278609    94***
201611072166    94***
201611283529    94***
201706149344    94***
201706300814    94***
M803667         94***
M804227         94***
M804767         94***
Name: Zipcode, dtype: object

In [35]:
building_permits_without_duplicates["Location"].apply(replace_by_infotype).head(10)

Permit Number
201505065519    String
201604195146    String
201605278609    String
201611072166    String
201611283529    String
201706149344    String
201706300814    String
M803667         String
M804227         String
M804767         String
Name: Location, dtype: object

### 1.4 Nunca modificar los datos crudos u originales

In [36]:
building_permits_without_duplicates.to_csv("building_permits-for-pandas.csv")